README.md
<h1><img align="right" src="./sticker.svg" width="30%" /> IntervalTrees</h1>

[![Latest Release](https://img.shields.io/github/release/BioJulia/IntervalTrees.jl.svg?style=flat-square)](https://github.com/BioJulia/IntervalTrees.jl/releases/latest)
[![MIT License](https://img.shields.io/badge/license-MIT-green.svg?style=flat-square)](https://github.com/BioJulia/IntervalTrees.jl/blob/master/LICENSE)
[![Stable documentation](https://img.shields.io/badge/docs-stable-blue.svg?style=flat-square)](https://biojulia.github.io/IntervalTrees.jl/stable)
![Lifecycle](https://img.shields.io/badge/lifecycle-stable-green.svg?style=flat-square)
[![Chat on Discord](https://img.shields.io/badge/discord-chat-blue.svg?style=flat-square&logo=discord&colorB=%237289DA)](https://discord.gg/z73YNFz)

## Description

IntervalTrees provides the type: `IntervalTree{K, V}`.
It implements an associative container mapping `(K, K)` pairs to to values of type `V`.
`K` may be any ordered type, but only pairs `(a, b)` where `a ≤ b` can be stored.
In other words they are associative contains that map intervals to values.


## Installation

Install IntervalTrees from the Julia REPL:

```julia
using Pkg
add("IntervalTrees")
# Pkg.add("IntervalTrees") for julia prior to v0.7
```

If you are interested in the cutting edge of the development, please check out
the master branch to try new features before release.


## Testing

IntervalTrees is tested against Julia `0.7` on Linux, OS X, and Windows.

| **PackageEvaluator** | **Latest Build Status** |
|:--------------------:|:-----------------------:|
| [![](http://pkg.julialang.org/badges/IntervalTrees_0.7.svg)](http://pkg.julialang.org/?pkg=IntervalTrees) | [![](https://travis-ci.org/BioJulia/IntervalTrees.jl.svg?branch=master)](https://travis-ci.org/BioJulia/IntervalTrees.jl) [![Build status](https://ci.appveyor.com/api/projects/status/3rwbgr40npp95upe/branch/master?svg=true)](https://ci.appveyor.com/project/BenJWard/intervaltrees-jl/branch/master) [![](https://codecov.io/gh/BioJulia/IntervalTrees.jl/branch/master/graph/badge.svg)](https://codecov.io/gh/BioJulia/IntervalTrees.jl) |


## Contributing

We appreciate contributions from users including reporting bugs, fixing
issues, improving performance and adding new features.

Take a look at the [CONTRIBUTING](CONTRIBUTING.md) file provided with
every BioJulia package package for detailed contributor and maintainer
guidelines.


### Financial contributions

We also welcome financial contributions in full transparency on our
[open collective](https://opencollective.com/biojulia).
Anyone can file an expense. If the expense makes sense for the development
of the community, it will be "merged" in the ledger of our open collective by
the core contributors and the person who filed the expense will be reimbursed.


# Backers & Sponsors

Thank you to all our backers and sponsors!

Love our work and community? [Become a backer](https://opencollective.com/biojulia#backer).

[![backers](https://opencollective.com/biojulia/backers.svg?width=890)](https://opencollective.com/biojulia#backers)

Does your company use BioJulia? Help keep BioJulia feature rich and healthy by
[sponsoring the project](https://opencollective.com/biojulia#sponsor)
Your logo will show up here with a link to your website.

[![](https://opencollective.com/biojulia/sponsor/0/avatar.svg)](https://opencollective.com/biojulia/sponsor/0/website)
[![](https://opencollective.com/biojulia/sponsor/1/avatar.svg)](https://opencollective.com/biojulia/sponsor/1/website)
[![](https://opencollective.com/biojulia/sponsor/2/avatar.svg)](https://opencollective.com/biojulia/sponsor/2/website)
[![](https://opencollective.com/biojulia/sponsor/3/avatar.svg)](https://opencollective.com/biojulia/sponsor/3/website)
[![](https://opencollective.com/biojulia/sponsor/4/avatar.svg)](https://opencollective.com/biojulia/sponsor/4/website)
[![](https://opencollective.com/biojulia/sponsor/5/avatar.svg)](https://opencollective.com/biojulia/sponsor/5/website)
[![](https://opencollective.com/biojulia/sponsor/6/avatar.svg)](https://opencollective.com/biojulia/sponsor/6/website)
[![](https://opencollective.com/biojulia/sponsor/7/avatar.svg)](https://opencollective.com/biojulia/sponsor/7/website)
[![](https://opencollective.com/biojulia/sponsor/8/avatar.svg)](https://opencollective.com/biojulia/sponsor/8/website)
[![](https://opencollective.com/biojulia/sponsor/9/avatar.svg)](https://opencollective.com/biojulia/sponsor/9/website)


## Questions?

If you have a question about contributing or using BioJulia software, come
on over and chat to us on [Discord](https://discord.gg/z73YNFz), or you can try the
[Bio category of the Julia discourse site](https://discourse.julialang.org/c/domain/bio).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2018-08-10
### Added
- Substantial changes supporting julia v0.7.0 and v1.0.0.
- Project files, including CHANGELOG, CODE_OF_CONDUCT, and CONTRIBUTING.

### Removed
- :fire: Support for julia v0.6.x.

## [0.4.1] - 2017-11-13
### Changed
- Fixed bugs in `firstintersection` that can lead to missed intersections.

## [0.4.0] - 2017-11-08
### Added
- Support for julia v0.6.

### Changed
- A more precise `issorted` check.

### Removed
- :fire: julia v0.5 support.

## [0.3.0] - 2017-06-23
### Added
- Optional filter predicates to `intersect` and `findfirst` functions.

## [0.2.1] - 2017-06-10
### Changed
- Faster `nextintersection!` function.
- Internal changes in behaviour.
- Fixed constructor syntax.

## [0.2.0] - 2017-03-21
### Added
- Support for julia v0.6.

### Removed
- :fire: Support for julia v0.4.

## [0.1.1] - 2017-02-22

[Unreleased]: https://github.com/BioJulia/IntervalTrees.jl/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/BioJulia/IntervalTrees.jl/compare/v0.4.1...v1.0.0
[0.4.1]: https://github.com/BioJulia/IntervalTrees.jl/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com/BioJulia/IntervalTrees.jl/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/BioJulia/IntervalTrees.jl/compare/v0.2.1...v0.3.0
[0.2.1]: https://github.com/BioJulia/IntervalTrees.jl/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/BioJulia/IntervalTrees.jl/compare/v0.1.1...v0.2.0
[0.1.1]: https://github.com/BioJulia/IntervalTrees.jl/tree/v0.1.1


test/test.jl

module TestIntervalTrees

using FactCheck
using IntervalTrees


# Generating random intervals
srand(12345)
const maxend = int(1e9)

function randinterval(minstart=1, maxend=maxend)
    a = rand(minstart:maxend)
    b = rand(a:maxend)
    return (a, b)
end


# Verify that internal node keys and maxend values are correct
function validkeys(t::IntervalTrees.IntervalBTree)
    return validkeys(t.root, (0, 0), (maxend, maxend))
end


function validkeys(node::IntervalTrees.InternalNode, minint, maxint)
    for child in node.children
        if child.maxend > node.maxend
            return false
        end
    end

    for i in 1:length(node.keys)
        k = node.keys[i]

        if IntervalTrees.minkey(node.children[i+1]) != k
            return false
        end

        if !validkeys(node.children[i], minint, k) ||
           !validkeys(node.children[i+1], k, maxint)
           return false
       end
    end
    return true
end


function validkeys(node::IntervalTrees.LeafNode, minint, maxint)
    for k in node.keys
        if !(minint <= k < maxint) || k[2] > node.maxend
            return false
        end
    end
    return true
end


# Verify that parent pointers are correct
function validparents(t::IntervalTrees.IntervalBTree)
    if !isa(t.root.parent, IntervalTrees.NullNode)
        return false
    end

    return validparents(t.root)
end


function validparents(node::IntervalTrees.InternalNode)
    for child in node.children
        if !validparents(child) || child.parent != node
            return false
        end
    end
    return true
end

function validparents(node::IntervalTrees.LeafNode)
    return true
end


# Verify that sibling/cousin pointers are correct
function validsiblings(t::IntervalTrees.IntervalBTree)
    if !isa(t.root.left, IntervalTrees.NullNode) ||
       !isa(t.root.right, IntervalTrees.NullNode)
       return false
   end

   # Do an in-order traversal, pushing nodes onto a stack indexed by
   # depth.
   tdepth = depth(t)
   nodestacks = [Array(IntervalTrees.Node, 0) for _ in 1:tdepth]

   function _visit(node::IntervalTrees.InternalNode, k)
       push!(nodestacks[k], node)
       for child in node.children
           _visit(child, k+1)
       end
   end

   function _visit(node::IntervalTrees.LeafNode, k)
       push!(nodestacks[k], node)
   end

   _visit(t.root, 1)

   for k in 1:tdepth
       stack = nodestacks[k]
       if isempty(stack)
           continue
       end

       if !isa(stack[1].left, IntervalTrees.NullNode) ||
          !isa(stack[end].right, IntervalTrees.NullNode)
           return false
       end

       if length(stack) > 1
           if stack[1].right != stack[2] || stack[end].left != stack[end-1]
               return false
           end
       end

       for i in 2:length(stack)-1
           if stack[i].left != stack[i-1] || stack[i].right != stack[i+1]
               return false
           end
       end
   end

   return true
end


facts("Searching") do
    t = IntervalTree{Int, Int}()
    @fact haskey(t, (1,2)) => false

    n = 10000
    maxend = 1000000
    intervals = [randinterval(1, maxend) for i in 1:n]

    context("true positives") do
        for (i, interval) in enumerate(intervals)
            t[interval] = i
        end

        shuffle!(intervals)
        @fact all([haskey(t, interval) for interval in intervals]) => true
    end

    context("true negatives") do
        @fact all([!haskey(t, interval)
                   for interval in [randinterval(maxend+1, 2 * maxend)
                                    for i in 1:n]]) => true
    end
end


facts("Intersecting") do
    t = IntervalTree{Int, Int}()
   # TODO
end


facts("Iteration") do
    t = IntervalTree{Int, Int}()
    @fact isempty(collect(t)) => true
    # TODO
end


facts("Insertion") do
    t = IntervalTrees.IntervalTree{Int, Int}()
    n = 100000

    context("random insertions") do
        for v in 1:n
            k = randinterval()
            t[k] = v
        end
        @fact issorted(collect(keys(t))) => true
        @fact validkeys(t) => true
        @fact validparents(t) => true
        @fact validsiblings(t) => true
    end

    context("ordered insertions") do
        for v in 1:n
            k = (v,rand(v:maxend))
            t[k] = v
        end
        @fact issorted(collect(keys(t))) => true
        @fact validkeys(t) => true
        @fact validparents(t) => true
        @fact validsiblings(t) => true
    end

    context("reverse ordered insertions") do
        for v in 1:n
            k = (maxend-v,rand((maxend-v):maxend))
            t[k] = v
        end
        @fact issorted(collect(keys(t))) => true
        @fact validkeys(t) => true
        @fact validparents(t) => true
        @fact validsiblings(t) => true
    end
end


facts("Deletion") do
    t = IntervalTrees.IntervalTree{Int, Int}()
    n = 100000

    # insert n random intervals
    intervals = [randinterval() for _ in 1:n]
    for (i, interval) in enumerate(intervals)
        t[interval] = i
    end
    shuffle!(intervals)

    # delete one
    interval = pop!(intervals)
    delete!(t, interval)
    @fact validkeys(t) => true
    @fact validparents(t) => true
    @fact validsiblings(t) => true
    @fact issorted(collect(keys(t))) => true
    @fact haskey(t, interval) => false
    @fact all(map(interval -> haskey(t, interval), intervals)) => true

    # delete a random 50%
    for interval in 1:(n-1)/2
        interval = pop!(intervals)
        delete!(t, interval)
    end
    @fact validkeys(t) => true
    @fact validparents(t) => true
    @fact validsiblings(t) => true
    @fact issorted(collect(keys(t))) => true
    @fact all(map(interval -> haskey(t, interval), intervals)) => true

    # delete the rest
    for interval in intervals
        delete!(t, interval)
    end
    @fact isempty(t) => true
    @fact validkeys(t) => true
    @fact validparents(t) => true
    @fact validsiblings(t) => true
end


end




test/benchmark.jl
if VERSION > v"0.7-"
    push!(LOAD_PATH, Base.NamedEnv("v0.7"))
    using InteractiveUtils
    using Random
    using Profile
end
versioninfo()

using IntervalTrees
using CodecZlib

function readgff3(input::IO, chrom::String)
    intervals = Interval{Int}[]
    for line in eachline(input)
        if startswith(line, "#")
            continue
        end
        values = split(line, '\t')
        if values[1] == chrom
            chromstart = parse(Int, values[4])
            chromend = parse(Int, values[5])
            push!(intervals, Interval(chromstart, chromend))
        end
    end
    sort!(intervals)
    return IntervalTree{Int,Interval{Int}}(intervals)
end

function readgff3(filepath::String, chrom::String)
    stream = open(filepath)
    if endswith(filepath, ".gz")
        stream = GzipDecompressorStream(stream)
    end
    try
        return readgff3(stream, chrom)
    finally
        close(stream)
    end
end

isdir("data") || mkdir("data")
cd("data")
filename="Danio_rerio.GRCz11.92.gff3.gz"
isfile(filename)||
    download("ftp://ftp.ensembl.org/pub/release-92/gff3/danio_rerio/Danio_rerio.GRCz11.92.gff3.gz", filename)
cd("..")

tree = readgff3("data/Danio_rerio.GRCz11.92.gff3.gz", "1")

println("random query")
function random_query(tree, queries)
    n = 0
    for q in queries
        for _ in intersect(tree, q)
            n += 1
        end
    end
    return n
end
srand(1234)
intervals = shuffle!(collect(tree))
for _ in 1:5
    @time random_query(tree, intervals)
end
if "-p" in ARGS
    @profile random_query(tree, intervals)
    Profile.print()
end


src/map.jl
# IntervalTree map operations

const IntervalMap{K, V} = IntervalTree{K, IntervalValue{K, V}}


function Base.setindex!(t::IntervalBTree{K, V, B}, value,
                        key::Tuple{Any, Any}) where {K, V, B}
    push!(t, V(key[1], key[2], value), true)
end


function Base.setindex!(t::IntervalBTree{K, V, B}, value,
                        key::Interval{K}) where {K, V, B}
    push!(t, V(key.first, key.last, value), true)
end


function Base.setindex!(t::IntervalBTree{K, V, B}, value,
                        first, last) where {K, V, B}
    push!(t, V(first, last, value), true)
end


function Base.getindex(t::IntervalBTree{K, V, B}, key::AbstractInterval{K}) where {K, V, B}
    return _getindex(t.root, key)
end


function _getindex(t::InternalNode{K, V, B}, key::AbstractInterval{K}) where {K, V, B}
    i = findidx(t, key)
    if 1 <= length(t) - 1 && key >= t.keys[i]
        return _getindex(t.children[i+1], key)
    else
        return _getindex(t.children[i], key)
    end
end


function _getindex(t::LeafNode{K, V, B}, key::AbstractInterval{K}) where {K, V, B}
    i = findidx(t, key)
    if 1 <= i <= length(t) && first(t.entries[i]) == first(key) &&
        last(t.entries[i]) == last(key)
        return t.entries[i]
    else
        error(KeyError((key.first, key.last)))
    end
end


function Base.get(t::IntervalBTree{K, V, B}, key::Tuple{Any, Any}, default) where {K, V, B}
    return _get(t.root, Interval{K}(key[1], key[2]), default)
end


function Base.get(t::IntervalBTree{K, V, B}, key::AbstractInterval{K}, default) where {K, V, B}
    return _get(t.root, key, default)
end


function _get(t::InternalNode{K, V, B}, key::AbstractInterval{K}, default) where {K, V, B}
    i = findidx(t, key)
    if 1 <= length(t) - 1 && key >= t.keys[i]
        return _get(t.children[i+1], key, default)
    else
        return _get(t.children[i], key, default)
    end
end


function _get(t::LeafNode{K, V, B}, key::AbstractInterval{K}, default) where {K, V, B}
    i = findidx(t, key)
    if 1 <= i <= length(t) &&
        first(t.entries[i]) == first(key) && last(t.entries[i]) == last(key)
        return t.entries[i]
    else
        return default
    end
end


function Base.get!(t::IntervalBTree{K, V, B}, key::Tuple{Any, Any}, default) where {K, V, B}
    return push!(t, V(key[1], key[2], default), true, false)
end


function Base.get!(t::IntervalBTree{K, V, B}, key::AbstractInterval{K}, default) where {K, V, B}
    return push!(t, V(key.first, key.last, default), true, false)
end


function Base.delete!(t::IntervalBTree{K, V, B}, first, last) where {K, V, B}
    return deletefirst!(t, first, last)
end


function Base.delete!(t::IntervalBTree{K, V, B}, key::Tuple{Any, Any}) where {K, V, B}
    return deletefirst!(t, key)
end


function Base.delete!(t::IntervalBTree{K, V, B}, key::Interval{K}) where {K, V, B}
    return deletefirst!(t, key)
end


struct IntervalKeyIterator{K, V, B}
    t::IntervalBTree{K, V, B}
end

Base.eltype(::Type{IntervalKeyIterator{K,V,B}}) where {K,V,B} = Interval{K}
Base.IteratorSize(::Type{IntervalKeyIterator{K,V,B}}) where {K,V,B} = Base.SizeUnknown()


function Base.keys(t::IntervalBTree)
    return IntervalKeyIterator(t)
end

function Base.iterate(
        it::IntervalKeyIterator{K, V, B},
        state::IntervalBTreeIteratorState{K, V, B}=iterinitstate(it)) where {K, V, B}
    if state.leaf === nothing || isempty(notnothing(state.leaf))
        return nothing
    end
    leaf = state.leaf
    key = Interval{K}(first(leaf.entries[state.i]), last(leaf.entries[state.i]))
    if state.i < length(leaf)
        state = IntervalBTreeIteratorState{K, V, B}(leaf, state.i + 1)
    else
        state = IntervalBTreeIteratorState{K, V, B}(leaf.right, 1)
    end
    return key, state
end

function iterinitstate(it::IntervalKeyIterator{K, V, B}) where {K, V, B}
    # traverse to the first leaf node
    node = it.t.root
    while !isa(node, LeafNode{K, V, B})
        node = node.children[1]
    end
    return IntervalBTreeIteratorState(node, 1)
end


struct IntervalValueIterator{K, V <: IntervalValue, B}
    t::IntervalBTree{K, V, B}
end

Base.eltype(::Type{IntervalValueIterator{K,V,B}}) where {K,V,B} = valtype(V)
Base.IteratorSize(::Type{IntervalValueIterator{K,V,B}}) where {K,V,B} = Base.SizeUnknown()


function Base.values(t::IntervalBTree{K, V, B}) where {K, V<:IntervalValue, B}
    return IntervalValueIterator(t)
end

function Base.iterate(
        it::IntervalValueIterator{K, V, B},
        state::IntervalBTreeIteratorState{K, V, B}=iterinitstate(it)) where {K, V, B}
    if state.leaf === nothing || isempty(notnothing(state.leaf))
        return nothing
    end
    leaf = state.leaf
    value = leaf.entries[state.i].value
    if state.i < length(leaf)
        state = IntervalBTreeIteratorState{K, V, B}(leaf, state.i + 1)
    else
        state = IntervalBTreeIteratorState{K, V, B}(leaf.right, 1)
    end
    return value, state
end

function iterinitstate(it::IntervalValueIterator{K, V, B}) where {K, V, B}
    # traverse to the first leaf node
    node = it.t.root
    while !isa(node, LeafNode{K, V, B})
        node = node.children[1]
    end
    return IntervalBTreeIteratorState(node, 1)
end


src/slice.jl
# This is very simple implementation of go-style slices. Nodes in the B-tree
# have fixed sized arrays that change when elements are inserted or deleted.
# We would like for those to behave as though they are resized without actually
# doing any reallocation.

mutable struct Slice{T, N} <: AbstractVector{T}
    data::Vector{T}
    n::Int # Number of stored elements

    function Slice{T,N}() where {T,N}
        return new{T,N}(Vector{T}(undef, N), 0)
    end
end

function Base.length(s::Slice)
    return s.n
end


function Base.size(s::Slice)
    return (s.n,)
end

function Base.getindex(s::Slice{T, N}, i::Integer) where {T, N}
    if 1 <= i <= s.n
        @inbounds x = s.data[i]
        return x
    else
        throw(BoundsError())
    end
end


@inline function unsafe_getindex(s::Slice{T, N}, i::Integer) where {T, N}
    @inbounds x = s.data[i]
    return x
end


function Base.setindex!(s::Slice{T, N}, value, i::Integer) where {T, N}
    if 1 <= i <= s.n
        @inbounds s.data[i] = value
        return value
    else
        throw(BoundsError())
    end
end


function Base.push!(s::Slice{T, N}, value) where {T, N}
    if s.n < N
        s.n += 1
        @inbounds s.data[s.n] = value
    else
        throw(BoundsError("Slice expanded past fixed size."))
    end
end


function Base.pop!(s::Slice{T, N}) where {T, N}
    if s.n > 0
        @inbounds x = s.data[s.n]
        s.n -= 1
        return x
    else
        throw(BoundsError())
    end
end


function Base.insert!(s::Slice{T, N}, i::Integer, value) where {T, N}
    if s.n < N && 1 <= i <= s.n + 1
        # TODO: This should work but won't. Fix this in Julia.
        #copy!(s.data, i+1, s.data, i, s.n - i + 1)
        if isbitstype(T)
            unsafe_copyto!(pointer(s.data, i+1), pointer(s.data, i), s.n - i + 1)
        else
            for k in 0:(s.n - i)
                @inbounds s.data[s.n+1-k] = s.data[s.n-k]
            end
        end
        @inbounds s.data[i] = value
        s.n += 1
        return s
    else
        throw(BoundsError())
    end
end


function Base.resize!(s::Slice{T, N}, n::Integer) where {T, N}
    if 1 <= n <= N
        s.n = n
        return s
    else
        throw(BoundsError())
    end
end


function Base.splice!(s::Slice{T, N}, i::Integer) where {T, N}
    if 1 <= i <= s.n
        @inbounds x = s.data[i]
        for j in i:s.n-1
            @inbounds s.data[j] = s.data[j + 1]
        end
        s.n -= 1
        return x
    else
        throw(BoundsError())
    end
end


function Base.searchsortedfirst(s::Slice, x)
    return searchsortedfirst(s.data, x, 1, s.n, Base.Order.Forward)
end



function slice_insert!(xs::Vector{T}, count::Integer, i::Integer, value::T) where T
    if count < length(xs) && 1 <= i <= count + 1
        if isbitstype(T)
            unsafe_copyto!(pointer(xs, i+1), pointer(xs, i), count - i + 1)
        else
            for k in 0:(count - i)
                @inbounds xs[count+1-k] = xs[count-k]
            end
        end
        @inbounds xs[i] = value
        count += 1
        return count
    else
        throw(BoundsError())
    end
end


function slice_splice!(xs::Vector{T}, count::Integer, i::Integer) where T
    if 1 <= i <= count
        @inbounds x = xs[i]
        for j in i:count-1
            @inbounds xs[j] = xs[j + 1]
        end
        count -= 1
        return x, count
    else
        throw(BoundsError())
    end
end


latest/search_index.js
var documenterSearchIndex = {"docs": [

{
    "location": "index.html#",
    "page": "Home",
    "title": "Home",
    "category": "page",
    "text": ""
},

{
    "location": "index.html#IntervalTrees-1",
    "page": "Home",
    "title": "IntervalTrees",
    "category": "section",
    "text": "(Image: Latest Release) (Image: MIT License) (Image: Stable documentation) (Image: Lifecycle) (Image: Chat on Discord)"
},

{
    "location": "index.html#Description-1",
    "page": "Home",
    "title": "Description",
    "category": "section",
    "text": "IntervalTrees provides the type: IntervalTree{K, V}. It implements an associative container mapping (K, K) pairs to to values of type V. K may be any ordered type, but only pairs (a, b) where a ≤ b can be stored. In other words they are associative contains that map intervals to values."
},

{
    "location": "index.html#Installation-1",
    "page": "Home",
    "title": "Installation",
    "category": "section",
    "text": "Install IntervalTrees from the Julia REPL:using Pkg\nadd(\"IntervalTrees\")\n# Pkg.add(\"IntervalTrees\") for julia prior to v0.7If you are interested in the cutting edge of the development, please check out the master branch to try new features before release."
},

{
    "location": "index.html#Testing-1",
    "page": "Home",
    "title": "Testing",
    "category": "section",
    "text": "IntervalTrees is tested against Julia 0.7 on Linux, OS X, and Windows.PackageEvaluator Latest Build Status\n(Image: ) (Image: ) (Image: Build status) (Image: )"
},

{
    "location": "index.html#Contributing-1",
    "page": "Home",
    "title": "Contributing",
    "category": "section",
    "text": "We appreciate contributions from users including reporting bugs, fixing issues, improving performance and adding new features.Take a look at the CONTRIBUTING file provided with every BioJulia package package for detailed contributor and maintainer guidelines."
},

{
    "location": "index.html#Financial-contributions-1",
    "page": "Home",
    "title": "Financial contributions",
    "category": "section",
    "text": "We also welcome financial contributions in full transparency on our open collective. Anyone can file an expense. If the expense makes sense for the development of the community, it will be \"merged\" in the ledger of our open collective by the core contributors and the person who filed the expense will be reimbursed."
},

{
    "location": "index.html#Backers-and-Sponsors-1",
    "page": "Home",
    "title": "Backers & Sponsors",
    "category": "section",
    "text": "Thank you to all our backers and sponsors!Love our work and community? Become a backer.(Image: backers)Does your company use BioJulia? Help keep BioJulia feature rich and healthy by sponsoring the project Your logo will show up here with a link to your website.(Image: ) (Image: ) (Image: ) (Image: ) (Image: ) (Image: ) (Image: ) (Image: ) (Image: ) (Image: )"
},

{
    "location": "index.html#Questions?-1",
    "page": "Home",
    "title": "Questions?",
    "category": "section",
    "text": "If you have a question about contributing or using BioJulia software, come on over and chat to us on Discord, or you can try the Bio category of the Julia discourse site."
},

{
    "location": "man/types.html#",
    "page": "Types",
    "title": "Types",
    "category": "page",
    "text": "CurrentModule = IntervalTrees"
},

{
    "location": "man/types.html#IntervalTrees.AbstractInterval",
    "page": "Types",
    "title": "IntervalTrees.AbstractInterval",
    "category": "type",
    "text": "Types deriving from AbstractInterval{T} must have a first and last function each returning a value of type T, and first(i) <= last(i) must always be true.\n\n\n\n\n\n"
},

{
    "location": "man/types.html#IntervalTrees.Interval",
    "page": "Types",
    "title": "IntervalTrees.Interval",
    "category": "type",
    "text": "A basic interval.\n\n\n\n\n\n"
},

{
    "location": "man/types.html#IntervalTrees.IntervalValue",
    "page": "Types",
    "title": "IntervalTrees.IntervalValue",
    "category": "type",
    "text": "An interval with some associated data.\n\n\n\n\n\n"
},

{
    "location": "man/types.html#Intervals-and-IntervalTrees-1",
    "page": "Types",
    "title": "Intervals & IntervalTrees",
    "category": "section",
    "text": "IntervalTrees exports an abstract type AbstractInterval{K}, in addition to two basic concrete interval types.AbstractInterval\nInterval\nIntervalValueIntervals in this package are always treated as end-inclusive, similar to the Julia Range type."
},

{
    "location": "man/types.html#IntervalTrees-1",
    "page": "Types",
    "title": "IntervalTrees",
    "category": "section",
    "text": "The basic data structure implemented in this package is IntervalTree{K, V}, which stores intervals of type V, that have start and end positions of type K.IntervalMap{K, V} is a typealias for IntervalTree{K, IntervalValue{K, V}} to simplify associating data of type V with intervals.Multiple data structures are refered to as \"interval trees\". What\'s implemented here is the data structure described in the Cormen, et al. \"Algorithms\" book, or what\'s refered to as an augmented tree in the wikipedia article. This sort of data structure is just an balanced search tree augmented with a field to keep track of the maximum interval end point in that node\'s subtree.Many operations over two or more sets of intervals can be most efficiently implemented by jointly iterating over the sets in order. For example, finding all the intersecting intervals in two sets S and T can be implemented similarly to the merge function in mergesort in O(n+m) time.Thus a general purpose data structure should be optimized for fast in-order iteration while efficiently supporting other operations like insertion, deletion, and single intersection tests. A B+-tree is nicely suited to the task. Since all the intervals and values are stored in contiguous memory in the leaf nodes, and the leaf nodes augmented with sibling pointers, in-order traversal is exceedingly efficient compared to other balanced search trees, while other operations are comparable in performance."
},

{
    "location": "man/insertion.html#",
    "page": "Creating IntervalTrees",
    "title": "Creating IntervalTrees",
    "category": "page",
    "text": ""
},

{
    "location": "man/insertion.html#Creating-IntervalTrees-and-inserting-intervals-1",
    "page": "Creating IntervalTrees",
    "title": "Creating IntervalTrees & inserting intervals",
    "category": "section",
    "text": "New intervals can be added to an IntervalTree with the push! function.xs = IntervalTree{Int, Interval{Int}}()\npush!(xs, Interval{Int}(500, 1000))A more efficient means of building the data structure by bulk insertion. If the intervals are knows up front and provided in a sorted array, an IntervalTree can be built extremely efficiently.intervals = Interval{Int}[]\n# construct a large array of intervals...\n\nsort!(intervals)\nxs = IntervalTree{Int, Interval{Int}}(intervals)"
},

{
    "location": "man/dict.html#",
    "page": "Dictionary Operations",
    "title": "Dictionary Operations",
    "category": "page",
    "text": ""
},

{
    "location": "man/dict.html#Standard-Dictionary-Operations-1",
    "page": "Dictionary Operations",
    "title": "Standard Dictionary Operations",
    "category": "section",
    "text": "IntervalTree implements all the standard dictionary operations. You can use it as an efficient way to map (K, K) tuples to values.using IntervalTrees\n\n# Create an interval tree mapping (Int, Int) intervals to Strings.\nxs = IntervalMap{Int, String}()\n\n# Insert values\nxs[(1,100)] = \"Low\"\nxs[(101,1000)] = \"Medium\"\nxs[(1001,10000)] = \"High\"\n\n# Search for values\nprintln(xs[(1001,10000)]) # prints \"High\"\n\n# Get a value, returning a default value if not found\nprintln(get(xs, (10001, 100000), \"Not found\")) # prints \"Not found\"\n\n# Set a value if it\'s not already present\nprintln(set(xs, (10001, 100000), \"Not found\"))\n\n# Delete values\ndelete!(xs, (1,100))\n"
},

{
    "location": "man/iteration.html#",
    "page": "Iterating over intervals",
    "title": "Iterating over intervals",
    "category": "page",
    "text": ""
},

{
    "location": "man/iteration.html#Iteration-1",
    "page": "Iterating over intervals",
    "title": "Iteration",
    "category": "section",
    "text": "As with dictionaries, key/value pairs can be iterated through efficiently.    for x in xs\n        println(\"Interval $(x.first), $(x.last) has value $(x.value)\")\n    endSome other iteration functions are provided:from(t::IntervalTree, query): Return an iterator thats iterates through every key/value pair with an end position >= to query.keys(t::IntervalTree): Return an iterator that iterates through every interval key in the tree.values(t::IntervalTree): Return an iterator that iterates through every value in the tree."
},

{
    "location": "man/intersection.html#",
    "page": "Intersection",
    "title": "Intersection",
    "category": "page",
    "text": ""
},

{
    "location": "man/intersection.html#Intersection-1",
    "page": "Intersection",
    "title": "Intersection",
    "category": "section",
    "text": "The primary thing an IntervalTree offers over a Dict is the ability to efficiently find intersections. IntervalTrees supports searching and iterating over intersections between two trees or between a tree and a single interval.intersect(t::IntervalTree, query::(Any, Any)): Return an iterator over every interval in t that intersects query.intersect(t1::IntervalTree, t2::IntervalTree): Return an iterator over every pair of intersecting entries (interval1, interval2), where interval1 is in t1 and interval2 is in t2.hasintersection(t::IntervalTree, position): Return true if position intersects some interval in t."
},

]}
